File: src/features/sync/subscriptionState.test.ts
// @vitest-environment jsdom
import { describe, expect, it } from 'vitest';

import { resolveLocalizedMessage } from '$shared/localization';

import { subscriptionStateMessage } from './subscriptionState';
import type { BillingStatusOutput } from './syncContract.generated';

function billing(overrides: Partial<BillingStatusOutput> = {}): BillingStatusOutput {
  return {
    entitled: true,
    state: 'active',
    graceUntil: null,
    storageQuotaBytes: 10_000_000_000,
    blobMaxBytes: 104_857_600,
    bytesUsed: 0,
    ...overrides,
  };
}

describe('subscriptionStateMessage', () => {
  it('says the subscription state in words, not the provider vocabulary', () => {
    const cases: Array<[string, string]> = [
      ['active', 'Active'],
      ['trialing', 'Trial'],
      ['canceled', 'Expired'],
      ['paused', 'Paused'],
      ['none', 'No subscription'],
    ];
    for (const [state, expected] of cases) {
      expect(resolveLocalizedMessage(subscriptionStateMessage(billing({ state })))).toBe(expected);
    }
  });

  it('tells a past-due account when sync stops, because that is what it can act on', () => {
    const graceUntil = new Date(Date.now() + 7.5 * 24 * 60 * 60 * 1000).toISOString();
    const text = resolveLocalizedMessage(
      subscriptionStateMessage(billing({ state: 'past_due', entitled: true, graceUntil })),
    );
    expect(text).toContain('Payment failed');
    expect(text).toContain('7 days');
  });

  it('still says payment failed when the server sent no grace date', () => {
    const text = resolveLocalizedMessage(
      subscriptionStateMessage(billing({ state: 'past_due', graceUntil: null })),
    );
    expect(text).toBe('Payment failed. Sync will pause.');
  });

  it('ignores an unparseable grace date rather than printing a broken one', () => {
    const text = resolveLocalizedMessage(
      subscriptionStateMessage(billing({ state: 'unpaid', graceUntil: 'not a date' })),
    );
    expect(text).toBe('Payment failed. Sync will pause.');
  });

  it('falls back to the entitlement for a state it has never seen', () => {
    expect(
      resolveLocalizedMessage(
        subscriptionStateMessage(billing({ state: 'quantum', entitled: true })),
      ),
    ).toBe('Active');
    expect(
      resolveLocalizedMessage(
        subscriptionStateMessage(billing({ state: 'quantum', entitled: false })),
      ),
    ).toBe('No subscription');
  });
});

File: apps/android/app/src/test/java/com/futo/notes/sync/hosted/SubscriptionStateTest.kt
package com.futo.notes.sync.hosted

import org.junit.Assert.assertEquals
import org.junit.Test
import uniffi.futo_notes_ffi.BillingStatus

/**
 * The subscription in words. The payment provider's vocabulary becomes a
 * sentence exactly here, and a state nobody recognised must still say something
 * true.
 */
class SubscriptionStateTest {
    /** Stands in for the relative-time renderer, so the wording is pinned. */
    private val relativeTime: (Long) -> String = { "in 4 days" }

    private fun billing(
        state: String,
        entitled: Boolean = true,
        graceUntil: String? = null,
    ) = BillingStatus(
        entitled = entitled,
        state = state,
        graceUntil = graceUntil,
        storageQuotaBytes = 10_000_000_000uL,
        blobMaxBytes = 104_857_600uL,
        bytesUsed = 4_210_688uL,
    )

    @Test
    fun `each state the provider sends has its own wording`() {
        val cases = listOf(
            "active" to "sync.hosted.account.state.active",
            "trialing" to "sync.hosted.account.state.trialing",
            "canceled" to "sync.hosted.account.state.expired",
            "paused" to "sync.hosted.account.state.paused",
            "incomplete" to "sync.hosted.account.state.incomplete",
            "incomplete_expired" to "sync.hosted.account.state.incomplete",
            "none" to "sync.hosted.account.state.none",
        )
        for ((state, path) in cases) {
            assertEquals(path, subscriptionStateMessage(billing(state), relativeTime).path)
        }
    }

    @Test
    fun `a failed payment says when sync will pause`() {
        val message = subscriptionStateMessage(
            billing("past_due", graceUntil = "2026-09-22T00:00:00Z"),
            relativeTime,
        )
        assertEquals("sync.hosted.account.state.pastDue", message.path)
        assertEquals("in 4 days", message.arguments["when"])
    }

    @Test
    fun `fractional seconds in the grace date are still a date`() {
        val message = subscriptionStateMessage(
            billing("unpaid", graceUntil = "2026-09-22T00:00:00.123Z"),
            relativeTime,
        )
        assertEquals("sync.hosted.account.state.pastDue", message.path)
    }

    @Test
    fun `a failed payment with no usable date still says payment failed`() {
        assertEquals(
            "sync.hosted.account.state.pastDueNoDate",
            subscriptionStateMessage(billing("past_due", graceUntil = null), relativeTime).path,
        )
        assertEquals(
            "sync.hosted.account.state.pastDueNoDate",
            subscriptionStateMessage(
                billing("past_due", graceUntil = "not a date"),
                relativeTime,
            ).path,
        )
    }

    @Test
    fun `a state nobody recognised falls back to the entitlement`() {
        assertEquals(
            "sync.hosted.account.state.active",
            subscriptionStateMessage(billing("something_new", entitled = true), relativeTime).path,
        )
        assertEquals(
            "sync.hosted.account.state.none",
            subscriptionStateMessage(billing("something_new", entitled = false), relativeTime).path,
        )
    }
}
